### PR TITLE
Propagate nullability from Kotlin metadata

### PR DESCRIPTION
**Pull request by volivan239**

### CODE REVIEW

File: jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt [hunk 1]
```diff
@@ -52,6 +64,44 @@ val JcMethod.kmConstructor: KmConstructor?
     get() =
         enclosingClass.kMetadata?.constructors?.firstOrNull { it.signature?.name == name && it.signature?.desc == description }
 
+val JcParameter.kmParameter: KmValueParameter?
+    get() {
+        method.kmFunction?.let {
+            // Shift needed to properly handle extension functions
+            val shift = if (it.receiverParameterType != null) 1 else 0
```
Comment (lehvolk) [inline jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt:71]: Could you please merge `main` branch and apply fix from there.

File: jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt [hunk 2]
```diff
@@ -44,6 +52,10 @@ val JcClassOrInterface.kMetadata: KotlinClassMetadata?
         return KotlinClassMetadata.read(kmHeader)
```
Comment (lehvolk) [inline jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt:52]: there is no guarantee that this code will not fail. Let's wrap into try-catch and write a warning in logs about it

File: jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt [hunk 3]
```diff
@@ -16,54 +16,103 @@
...
+import org.utbot.jcdb.api.ext.isNullable
 
-sealed class JvmType {
+sealed class JvmType(val isNullable: Boolean) {
 
     abstract val displayName: String
 
+    abstract fun setNullability(nullable: Boolean): JvmType
+
+    open fun relaxWithKmType(kmType: KmType): JvmType =
```
Comment (lehvolk) [inline jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt:29]: For sure this is shouldn't be interface method of JvmType

File: jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt [hunk 4]
```diff
@@ -16,54 +16,103 @@
...
+import kotlinx.metadata.KmType
 import org.utbot.jcdb.api.PredefinedPrimitives
+import org.utbot.jcdb.api.ext.isNullable
 
-sealed class JvmType {
+sealed class JvmType(val isNullable: Boolean) {
 
     abstract val displayName: String
 
+    abstract fun setNullability(nullable: Boolean): JvmType
```
Comment (lehvolk) [inline jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt:27]: This method looks like setter but should not be setter. Lets stay with `immutable` pattern naming and do something like for class level api: `notNullable()` or something similar

File: jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt [hunk 5]
```diff
@@ -72,6 +121,11 @@ open class JvmTypeVariable(val symbol: String) : JvmType() {
     override val displayName: String
         get() = symbol
 
+    override fun setNullability(nullable: Boolean): JvmType =
+        JvmTypeVariable(symbol, nullable).also {
+            it.declaration = declaration
+        }
+
     override fun equals(other: Any?): Boolean {
```
Comment (lehvolk) [inline jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt:129]: Am I right that both types with `nullable = true` and `nullable=false` will be equal?

**Review by lehvolk — CHANGES_REQUESTED** [review]

File: jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt (hunk 1 above)
Comment (volivan239) [reply jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt]: Done

File: jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt (hunk 2 above)
Comment (volivan239) [reply jcdb-api/src/main/kotlin/org/utbot/jcdb/api/ext/KMetadataExtension.kt]: Done

File: jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt (hunk 3 above)
Comment (volivan239) [reply jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt]: Rewrote this logic using extension functions

File: jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt (hunk 5 above)
Comment (volivan239) [reply jcdb-core/src/main/kotlin/org/utbot/jcdb/impl/types/signature/JvmTypes.kt]: Deleted `equals` and `hashCode` completely as legacy

**Pull request merged**